### PR TITLE
feat(server): secure-by-default bind-auth gate — refuse unauthenticated non-loopback bind

### PR DESCRIPTION
**Pull request by youhaowei**
Guard `createDashframeServer` so a non-loopback bind without an `authToken` fails fast with a clear error. Loopback (127.x / ::1 / localhost) remains allowed without a token for local dev and Electron. The `insecure` option mirrors the CLI's `--insecure` escape hatch for deliberate opt-out.

## Changes

- `apps/server/src/app.ts`: Add `isLoopbackHost` helper; add bind-auth guard at the top of `createDashframeServer` that throws before any I/O when hostname is non-loopback and `authToken` is absent (unless `insecure: true`). Add `insecure?: boolean` to `DashframeServerOptions`. Update `authToken` JSDoc to document the new requirement.
- `apps/server/src/app.test.ts`: Add `bind-auth gate` describe block with three contract tests: (a) non-loopback + no token → throws; (b) non-loopback + token → ok; (c) loopback + no token → ok.

## Screenshots

No UI change.

## Verification

- `bun run --cwd apps/server test`: 168 tests passed (165 pre-existing + 3 new contract tests). All three gate cases are covered and green.
- `bun run --cwd apps/server lint`: clean.
- Rebased on `origin/main` (ecfa1f48) before push.

<details open><summary><h3>Greptile Summary</h3></summary>

This PR introduces a secure-by-default bind-auth gate to `createDashframeServer`: non-loopback binds now require either an `authToken` or an explicit `insecure: true` opt-out, failing fast before any socket is opened. The gate logic is extracted into the exported `assertBindAuthorized` helper so the security-critical "token allows non-loopback" branch can be tested against a real non-loopback address (`0.0.0.0`) without actually binding a socket in CI.

- `assertBindAuthorized` correctly handles all branches: refuse non-loopback+no-token, allow non-loopback+token, allow loopback+no-token, allow non-loopback+insecure (with a `console.warn`). `isLoopbackHost` covers the full 127.0.0.0/8 block, `::1`, `localhost`, and the `undefined` default.
- Five tests cover every gate branch plus an end-to-end wiring check through `createDashframeServer` itself, which verifies the gate fires before any real I/O.
- The `DashframeServerOptions` interface and JSDoc are updated to document the new requirement clearly.
</details>

<details><summary><h3>Confidence Score: 5/5</h3></summary>

Safe to merge — the gate is a straightforward early-return guard with no I/O side-effects on the reject path, and all four allow/deny branches are directly exercised by tests.

The bind-auth gate is a simple, well-scoped addition: it throws before any socket is opened, the logic is deterministic, and the extraction of assertBindAuthorized ensures every branch (refuse, allow-with-token, allow-loopback, insecure opt-out) is covered by direct unit tests plus an end-to-end wiring test. No existing behavior is changed for the common loopback case.

No files require special attention.
</details>

<details><summary><h3>Important Files Changed</h3></summary>

| Filename | Overview |
|----------|----------|
| apps/server/src/app.ts | Adds isLoopbackHost helper and exported assertBindAuthorized gate; gate is correctly invoked at the top of createDashframeServer before any socket I/O. Logic, edge cases (127.x/8, ::1, localhost, undefined default), and interface/JSDoc updates all look correct. |
| apps/server/src/app.test.ts | Five new tests in the bind-auth gate describe block cover all four assertBindAuthorized branches plus an end-to-end wiring check through createDashframeServer. Tests correctly use 0.0.0.0 for non-loopback branches to exercise the real allow/deny logic without binding a socket. |

</details>

</details>

<details><summary><h3>Flowchart</h3></summary>

```mermaid
%%{init: {'theme': 'neutral'}}%%
flowchart TD
    A["createDashframeServer(opts)"] --> B["Resolve hostname\n(opts.hostname ?? '127.0.0.1')"]
    B --> C["assertBindAuthorized(hostname, authToken, insecure)"]
    C --> D{"isLoopbackHost(hostname)?"}
    D -- Yes --> G["✅ Gate passes\n(loopback, no token needed)"]
    D -- No --> E{"authToken present?"}
    E -- Yes --> F["✅ Gate passes\n(token satisfies non-loopback)"]
    E -- No --> H{"insecure: true?"}
    H -- Yes --> I["console.warn\n⚠️ insecure non-loopback"]
    I --> J["✅ Gate passes\n(deliberate opt-out)"]
    H -- No --> K["❌ throw Error\n'refusing to bind … without an auth token'"]
    G --> L["Continue: bind socket,\nstart HTTP+WS server"]
    F --> L
    J --> L
```
</details>

<sub>Reviews (2): Last reviewed commit: ["test(server): exercise the token-allows-..."](https://github.com/youhaowei/dashframe/commit/625f6d8c706f4e13b49710561393644b4dd689a8) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=37462060)</sub>

### CODE REVIEW

File: apps/server/src/app.test.ts [hunk 1]
```diff
@@ -62,6 +62,76 @@ function waitForWsAuth(
...
+        db: project.db,
+        hostname: "127.0.0.1", // loopback address to avoid actually binding to 0.0.0.0 in CI
+        authToken: "a-valid-token",
+      });
+      expect(server.port).toBeGreaterThan(0);
+    } finally {
+      server?.stop();
+      await project.close();
+    }
+  });
```
[the remark this answers is not in the document]
Comment (youhaowei) [reply apps/server/src/app.test.ts]: Fixed in 625f6d8c. You're right — the test passed via the loopback short-circuit and never exercised the token-allows-non-loopback branch.

Rather than relabel the test, I extracted the gate decision into an exported pure function `assertBindAuthorized({ hostname, authToken, insecure })` (the same function `createDashframeServer` now calls before any socket bind) and test the branches directly against a genuinely non-loopback host (`0.0.0.0`) — no socket, no DB:
- `non-loopback + token → allowed` now runs with `hostname: "0.0.0.0"`, so `!isLoopbackHost` is true and the bind is permitted *because of the token*.
- I verified it's no longer hollow: temporarily removing `&& !opts.authToken` from the gate makes exactly this test fail, then restored.

Also kept `non-loopback + no token → throws`, `loopback + no token → ok`, added `insecure` opt-out, and an end-to-end factory test for the deny path.

File: apps/server/src/app.test.ts [hunk 2]
```diff
@@ -62,6 +62,76 @@ function waitForWsAuth(
   });
 }
 
+describe("bind-auth gate", () => {
+  /**
+   * Secure-by-default: the factory must refuse to start when both conditions
+   * are true: (a) the bind host is non-loopback, AND (b) no authToken is set.
+   * These tests prove the three contract cases without opening a real DB or
+   * network socket — the guard fires before any I/O.
+   */
```
[the remark this answers is not in the document]
Comment (youhaowei) [reply apps/server/src/app.test.ts:77]: Fixed in 625f6d8c. The block comment now accurately states that the gate tests are I/O-free (they call the pure `assertBindAuthorized`, no DB or socket), while the end-to-end factory test does open a real project DB. The misleading "all three tests open no real DB" claim is gone.

File: apps/server/src/app.test.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (youhaowei) [reply apps/server/src/app.test.ts]: Fixed in 625f6d8c. Agreed — `hostname: "127.0.0.1"` is loopback, so the test passed via the loopback-allow path and never verified that the token permits a non-loopback bind.

The gate decision is now an exported pure function `assertBindAuthorized`, and the allow-path test uses a genuinely non-loopback host (`0.0.0.0`) so the token branch is actually exercised with no real socket bind in CI. I confirmed the test fails if the token-allow branch is removed, then restored it.

**Comment by youhaowei** [comment]
Tracked internally as **YW-136** — https://linear.app/youhaowei/issue/YW-136

(Linking via comment since the branch name predates the ticket-id convention; the canonical link is the Linear identifier above. Future PRs will carry the id in the branch/title.)
